// react-native-stopwatch/src/button.js
import React, { Component } from 'react';
import {
    TouchableHighlight,
    StyleSheet,
    View,
    Text
} from 'react-native';

class Button extends Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <TouchableHighlight underlayColor="#e1e1e1"
                                disabled={this.props.disabled}
                                onPress={this.props.onPress}
                                style={this.props.disabled ? styles.disabledButton : styles.button}>
                <Text style={this.props.disabled ? styles.disabledText : this.props.textStyle}>
                    {this.props.text}
                </Text>
            </TouchableHighlight>
        );
    }
}

const styles = StyleSheet.create({
    disabledButton: {
        height: 75,
        width: 75,
        borderRadius: 50,
        justifyContent: 'space-around',
        alignItems: 'center',
        backgroundColor: 'rgb(252, 252, 252)'
    },
    disabledText: {
        color: 'rgb(130, 130, 130)'
    },
    button: {
        height: 75,
        width: 75,
        borderRadius: 50,
        justifyContent: 'space-around',
        alignItems: 'center',
        backgroundColor: '#fff'
    }
});

export default Button;
